src/Common/src/TypeSystem/IL/Stubs/CalliMarshallingMethodThunk.cs

```diff
@@ -0,0 +1,106 @@
+
+using System;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Interop;
+using Debug = System.Diagnostics.Debug;
+using Internal.TypeSystem.Ecma;
+
+namespace Internal.IL.Stubs
+{
+    /// <summary>
+    /// Thunk to marshal delegate parameters and invoke the appropriate delegate function pointer
+    /// </summary>
+    public partial class CalliMarshallingMethodThunk : ILStubMethod
+    {
+        private readonly MethodSignature _targetSignature;
+        private readonly InteropStateManager _interopStateManager;
+        private readonly TypeDesc _owningType;
+
+        private MethodSignature _signature;
+
+        public CalliMarshallingMethodThunk(MethodSignature targetSignature, TypeDesc owningType,
+                InteropStateManager interopStateManager)
+        {
+            _targetSignature = targetSignature;
+            _owningType = owningType;
+            _interopStateManager = interopStateManager;
+        }
+
+        public MethodSignature TargetSignature
+        {
+            get
+            {
+                return _targetSignature;
+            }
+        }
+
+        public override TypeSystemContext Context
+        {
+            get
+            {
+                return _owningType.Context;
+            }
+        }
+
+        public override TypeDesc OwningType
+        {
+            get
+            {
+                return _owningType;
+            }
+        }
+
+        public override MethodSignature Signature
+        {
+            get
+            {
+                if (_signature == null)
+                {
+                    // Prepend fnptr argument to the signature
+                    TypeDesc[] parameterTypes = new TypeDesc[_targetSignature.Length + 1];
+
+                    parameterTypes[0] = Context.GetWellKnownType(WellKnownType.IntPtr);
+                    for (int i = 0; i < _targetSignature.Length; i++)
+                        parameterTypes[i + 1] = _targetSignature[i];
+
+                    _signature = new MethodSignature(MethodSignatureFlags.Static, 0, _targetSignature.ReturnType, parameterTypes);
+                }
+                return _signature;
+            }
+        }
+
+        public override string Name
+        {
+            get
+            {
+                return "CalliMarshallingMethodThunk";
+            }
+        }
+
+        public override MethodIL EmitIL()
+        {
+            // TODO: Emit marshaling stub
+            // return PInvokeILEmitter.EmitIL(this, default(PInvokeILEmitterConfiguration), _interopStateManager);
+
+            bool blittable = _targetSignature.ReturnType.IsVoid || MarshalHelpers.IsBlittableType(_targetSignature.ReturnType);
+            for (int i = 0; i < _targetSignature.Length; i++)
+                blittable = blittable && MarshalHelpers.IsBlittableType(_targetSignature[i]);
+            if (!blittable)
+                throw new NotSupportedException("PInvoke Calli with non-blittable type");
```

Comment:
Depending on how long term this code is expected to stay - maybe this could generate a throwing body. I think throwing this exception type here will lead to a compilation failure.